torch/distributed/fsdp/fully_sharded_data_parallel.py

```diff
@@ -859,10 +861,23 @@ def summon_full_params(
                 constructor and ``offload_to_cpu=False`` to this method.
                 (Default: ``False``)
         """
-        with _unshard_params(
-            module, recurse, writeback, rank0_only, offload_to_cpu, with_grads
-        ):
-            yield
+        try:
+            module_iter = iter(module)
+            with contextlib.ExitStack() as stack:
+                unshard_contexts = [
+                    _unshard_params(
+                        m, recurse, writeback, rank0_only, offload_to_cpu, with_grads
+                    )
+                    for m in module_iter
+                ]
+                for mgr in unshard_contexts:
+                    stack.enter_context(mgr)
+                yield
+        except TypeError:
+            with _unshard_params(
+                module, recurse, writeback, rank0_only, offload_to_cpu, with_grads
+            ):
```

Comment:
@fegin yeah that makes sense. Since some modules are iterable, I'm planning to more explicitly separate the 2 codepaths by enforcing a `tuple` of modules has to be passed in if we want to do multiple summon.